okhttp/src/main/kotlin/com/connectrpc/okhttp/OkHttpStream.kt

```diff
@@ -195,7 +210,15 @@ internal class PipeRequestBody(
     override fun contentType() = contentType
 
     override fun writeTo(sink: BufferedSink) {
-        pipe.fold(sink)
+        pipeLock.withLock {
+            pipe.fold(sink)
+        }
+        if (!duplex) {
+            // For non-duplex request bodies, okhttp3
+            // expects this method to return only when
+            // the request body is complete.
+            closed.await()
```

Comment:
Why is this necessary? I don't see this logic in any of the other implementations of RequestBody or documented in the contract for writeTo.

Reply:
That's because other implementations of `RequestBody` aren't pipes, where they are awaiting data from another writer thread.

It is in fact documented: https://github.com/square/okhttp/blob/ee9ebba20a39e6e4cdd78cb74f488728e49de942/okhttp/src/main/kotlin/okhttp3/RequestBody.kt#L71-L78

Basically, `writeTo` is expected to write the entire request body to the given sink. But what was here before, just `pipe.fold()` does not do that. It only copies any data to the sink that has already been written and buffered in the pipe and then returns. So to make sure that the entire request body is written, we must wait for the writer to signal the end via closing.

Reply:
Sadly, I never saw that doc. So figuring this out was a bear.

I first added tracing to the conformance test runner, so I could see the full HTTP request and response on failures. That's how I saw that sometimes the server saw an empty body with zero requests (and less often, a non-empty body with too few requests).

That's when I started tracking down the calls to `send` to figure out what was failing. That resulted in the first commit in this PR (correctly propagating failures in a `Result<Unit>` up to the caller and then checking the result and throwing on failure in the conformance client). That allowed me to see that calls to `stream.send` were failing with a "closed" exception.

So then I had to track down where the heck it was being closed. (You'll see a bunch of tracing added in #212, including ability to log stack traces, which helped me find it.) And it was being called from [here](https://github.com/square/okhttp/blob/ee9ebba20a39e6e4cdd78cb74f488728e49de942/okhttp/src/main/kotlin/okhttp3/internal/http/CallServerInterceptor.kt#L63).

That's when I realized the crux of the issue -- the body had to be _completely_ written here before returning.

Another formulation I considered here was something like so:
```kotlin
if (duplex) {
    pipe.fold(sink)
    return
}
sink.writeAll(pipe.source)
```
But that ends up buffering in the pipe and needing a little more overhead on each write, to go through the pipe's sink-to-source exchange, from the writer thread to this one. Whereas `pipe.fold` allows subsequent writes to go directly to the new destination sink. So it seemed better to keep using `pipe.fold`. 🤷